File: 2022/go/day12/day12.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day12.txt")
	height, start, end := readMap(string(input))

	part1 := shortestPath(height, map[xy]int{}, visit{start, 0},
		func(dh int) bool { return dh <= 1 },
		func(p xy) bool { return end == p })

	part2 := shortestPath(height, map[xy]int{}, visit{end, 0},
		func(dh int) bool { return dh >= -1 },
		func(p xy) bool { return height[p] == 'a' })

	fmt.Printf("part1: %v\npart2: %v\n", part1, part2)
}

func shortestPath(height map[xy]int, visits map[xy]int, loc visit, ok func(d int) bool, target func(xy) bool) int {
	min := 9999
	if target(loc.xy) {
		return loc.distance
	} else if old, visited := visits[loc.xy]; visited && old <= loc.distance {
		return min
	}
	visits[loc.xy] = loc.distance
	for _, m := range []xy{{-1, 0}, {1, 0}, {0, -1}, {0, 1}} {
		next := visit{xy{loc.x + m.x, loc.y + m.y}, loc.distance + 1}
		if h, inside := height[next.xy]; inside {
			deltaH := h - height[loc.xy]
			if !ok(deltaH) {
				continue
			}
			d := shortestPath(height, visits, next, ok, target)
			if d < min {
				min = d
			}
		}
	}
	return min
}

func readMap(input string) (height map[xy]int, start, end xy) {
	rows := strings.Split(input, "\n")
	height = map[xy]int{}
	for y, row := range rows {
		for x, c := range row {
			if c == 'S' {
				start = xy{x, y}
				c = 'a'
			} else if c == 'E' {
				end = xy{x, y}
				c = 'z'
			}
			height[xy{x, y}] = int(c)
		}
	}
	return height, start, end
}

type visit struct {
	xy
	distance int
}

type xy struct {
	x, y int
}

File: 2022/java/Day12.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

class Day12 {

    public static void main(String ... args) throws Exception {
        var map = readMap(Files.readString(Path.of("../input/day12.txt")));
        XY start = map.keySet().stream().filter(k -> map.get(k) == 'S').findFirst().get();
        XY end = map.keySet().stream().filter(k -> map.get(k) == 'E').findFirst().get();
        map.put(start, (byte) 'a');
        map.put(end, (byte) 'z');

        System.out.printf("part1: %d\n",
            shortestPath(map, new HashMap<>(), start, dh -> dh <= 1, p -> p.equals(end)));
        System.out.printf("part2: %d\n",
            shortestPath(map, new HashMap<>(), end, dh -> dh >= -1, p -> map.get(p) == 'a'));
    }
 
    static int shortestPath(Map<XY,Byte> map, Map<XY,Integer> visits, XY start,
         Function<Byte,Boolean> ok, Function<XY,Boolean> target)  {        
        var visited = new HashMap<XY,Integer>();
        var toVisit = new LinkedList<Visit>();
        toVisit.add(new Visit(start, 0));
        while(!toVisit.isEmpty()) {
            var c = toVisit.removeFirst();
            if (!visited.containsKey(c.xy) || visited.get(c.xy) > c.distance) { // not visited or longer distance
                visited.put(c.xy, c.distance); // visit
                moves.stream().map(m -> c.move(m)) // next location
                    .filter(n -> map.containsKey(n.xy)) // is inside map
                    .filter(n -> ok.apply((byte) (map.get(n.xy) - map.get(c.xy)))) // ok to climb
                    .forEach(n -> toVisit.add(n));  // add to visit list
            }
        }

        return visited.keySet().stream()
            .filter(p -> target.apply(p)) // match or target
            .map(p -> visited.get(p)) // map to distance
            .sorted().findFirst().get(); // get minimum
    }

    static Map<XY,Byte> readMap(String input) {
        Map<XY,Byte> map = new HashMap<>();
        var lines = input.lines().toList();
        for (int y = 0; y < lines.size(); y++) {
            for (int x = 0; x < lines.get(y).length(); x++) {
                map.put(new XY(x,y), (byte) lines.get(y).charAt(x));
            }
        }
        return map;
    }

    record Visit(XY xy, int distance) {
        Visit move(XY v) { return new Visit(new XY(xy.x+v.x, xy.y+v.y), distance + 1); }
    }

    record XY(int x, int y) {}

    static List<XY> moves = List.of(new XY(1,0), new XY(-1,0), new XY(0,1), new XY(0,-1));
}